### PR TITLE
ENH: scipy.sparse.coo.coo_matrix.__init__: less memory needed

### PR DESCRIPTION
**Pull request by jor-**
During the creation of a coo matrix with row, column and data sequences the row and column arrays are created twice. They only have to be created once with this enhancement. As a consequence, memory and execution time can be saved. This is crucial for large amounts of data.

### CODE REVIEW

File: scipy/sparse/coo.py [hunk 1]
```diff
@@ -137,16 +137,13 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
                 except TypeError:
                     raise TypeError('invalid input format')
 
-                self.row = np.array(ij[0], copy=copy)
-                self.col = np.array(ij[1], copy=copy)
-                self.data = np.array(obj, copy=copy)
-
+                row, col = ij
```
Comment (larsmans) [inline scipy/sparse/coo.py]: Are you sure this is safe? I would have moved the `dtype=` stuff here. Now if a list is passed in, we get a slow Python `max` call in the validation code below.

**Comment by larsmans** [comment]
Agree with the need to trim down on copies. Have you profiled the speedup?

File: scipy/sparse/coo.py (hunk 1 above)
Comment (pv) [inline scipy/sparse/coo.py]: Better fix would be to just replace the two `astype` calls with
`self.row = np.asarray(self.row, dtype=idx_dtype)`, as was done in also
some other places.

**Comment by argriffing** [comment]
> Agree with the need to trim down on copies. Have you profiled the speedup?

Here the [peakmem](http://asv.readthedocs.org/en/latest/writing_benchmarks.html?highlight=peakmem#peak-memory) feature could be helpful.

File: scipy/sparse/coo.py (hunk 1 above)
Comment (jor-) [inline scipy/sparse/coo.py]: The replacement of the two `astype` calls with `asarray` calls does not prevent the creation of two arrays for each of `row` and `column` if `idx_dtype` is not the default `int` type.

**Comment by jor-** [comment]
Yes, the slower Python `max` call is a drawback.

The main advantage is that an array with the default `int` type (at my machine `int64`) is not generated in advance. If the necessary `int` type is lower (e.g. `int16`), a second array would be generated with the lower `int` type. This requires more memory.  Due to memory limits, this might make the use of matrices with many entries impossible.

It should be noted that the execution time might only be higher if no shape was passed due to the Python `max` call. If a shape is passed, the execution time is lower because only one and not two arrays are created for `row` and `column`. The memory usage is lower in both cases.

You have to decide whether to put more emphasis on lower memory usage or a small execution time here.

I have not profiled the exact speedup and memory usage yet, but I can do so in the next days if needed.

**Comment by larsmans** [comment]
But you're converting the `row` and `col` arrays anyway if the dtype is wrong. I would do that first, then check the shape, then store the arrays on the object. In this case, you only waste time if the shape check fails, and you never do more than one conversion/copy.

Profiling output is not strictly required, I'm just curious.

**Comment by jor-** [comment]
If the indices are stored in lists, in the old version it can happen that they are copied to arrays in line 141 and 142 and afterwards converted to another arrays in line 157 and 158. Or am I wrong?

**Comment by larsmans** [comment]
No, you're right, but you can do

    row = np.asarray(row, copy=copy, dtype=idx_type)

and similarly for `col`, then check the shape. So just do *one* conversion/copy, before the check.

**Comment by jor-** [comment]
The problem is that you need to know the `shape` to calculate `idx_dtype`.

**Comment by larsmans** [comment]
Oh, right. Excuse me.

But in that case I would replace `max` with `np.max`. That should work on anything that can be converted to an array, and it's orders of magnitude faster if it gets the expected input:

```
>>> a = np.arange(100000)
>>> %timeit max(a)
100 loops, best of 3: 7.25 ms per loop
>>> %timeit np.max(a)
10000 loops, best of 3: 92.5 µs per loop
```

**Comment by jor-** [comment]
Okay, I have change pythons `max` functions to numpys `max` functions.

**Comment by larsmans** [comment]
LGTM. Will merge if no-one voices objections.

**Comment by larsmans** [comment]
Merged. Thanks @jor-!

**Comment by jor-** [comment]
Fine. Thanks @larsmans for his input.

**Pull request merged**